### PR TITLE
Kh/forward references mirror nullable

### PR DESCRIPTION
**Pull request by Khatskevich**

### CODE REVIEW

File: avro_schema/frontend.lua [hunk 1]
```diff
@@ -389,7 +393,8 @@ copy_schema = function(schema, ns, scope, open_rec, options)
                 local symbolmap = {}
                 for symbolno, symbol in ipairs(xsymbols) do
                     symbol = tostring(symbol)
-                    if not validname(symbol) and not options.utf8_enums then
+                    if not validname(symbol)
+                        and not context.options.utf8_enums then
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Indent.

File: avro_schema/frontend.lua [hunk 2]
```diff
@@ -209,27 +211,29 @@ local function copy_fields(from, to, fields)
...
--- [open_rec] a set consisting of the current record + parent records;
---            it is used to reject records containing themselves
--- [options]  options table, contains:
---             - preserve_in_ast: names of attrs which should not be deleted
-copy_schema = function(schema, ns, scope, open_rec, options)
+-- context : table; stores necessary for parsing variables
+--   options: options passed by a user, contains:
+--     preserve_in_ast: names of attrs which should not be deleted
+--     utf8_enums: allow utf8 enum items
+--   scope: a dictionary of named types (ocasionally used for unnamed too)
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: ocasionally → occasionally

File: avro_schema/frontend.lua [hunk 3]
```diff
@@ -508,7 +547,7 @@ copy_schema_location_info = function()
             if node.type == nil then -- union
                 insert(res, '<union>')
                 if i <= top + 1 then
-                    local _, next_node = debug.getlocal(i - 1, 6)
+                    local _, next_node = debug.getlocal(i - 1, 5)
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Should be in the previous commit.

File: avro_schema/frontend.lua [hunk 4]
```diff
@@ -135,6 +141,54 @@ local copy_schema
...
+    context.defined[typeid] = DEFINITION.DEFINED
+    return xtype
+end
+
+-- get type
+local function scope_get_type(context, typeid, nullable)
+    local scope = context.scope
+    local _, typeid = extract_nullable(typeid)
+    local full_type = nullable and typeid .. "*" or typeid
+    if scope[full_type] then
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: If on REFERENCED | DEFINED or assert.

File: avro_schema/frontend.lua [hunk 5]
```diff
@@ -135,6 +141,54 @@ local copy_schema
...
 
+-- add new type definition to scope
+local function scope_add_type(context, typeid, xtype)
+    local scope = context.scope
+    local _, typeid = extract_nullable(typeid)
+    xtype.nullable = nil
+    if context.defined[typeid] == DEFINITION.DEFINED then
+        copy_schema_error('Type name already defined: %s', typeid)
+    end
+    -- make the type always store name in canonical form
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Maybe note that it is need only for 'DEFINED' types, but not 'REFERENCE' ones. It helps self-check for code correctness for future patches.

File: avro_schema/frontend.lua [hunk 6]
```diff
@@ -372,13 +426,12 @@ copy_schema = function(schema, context, ns, open_rec)
                 end
                 dcache[res] = fieldmap
                 open_rec[res] = nil
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Please, cherry-pick recursion ban by name into this commit.

File: avro_schema/frontend.lua (hunk 1 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: what is wrong with the indent?

File: avro_schema/frontend.lua (hunk 2 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: fixed

File: avro_schema/frontend.lua [hunk 7]
```diff
@@ -135,6 +141,54 @@ local copy_schema
...
+    if context.defined[typeid] == DEFINITION.DEFINED then
+        copy_schema_error('Type name already defined: %s', typeid)
+    end
+    -- make the type always store name in canonical form
+    xtype.name = typeid
+    -- In case of forward_reference it is necessary to check if this type is
+    -- used already somewhere, and reuse the same table for the type.
+    if context.defined[typeid] == DEFINITION.REFERENCED then
+        local deferred_def = scope[typeid]
+        assert(deferred_def)
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/frontend.lua]: added

File: avro_schema/frontend.lua [hunk 8]
```diff
@@ -548,9 +587,49 @@ copy_schema_error = function(fmt, ...)
...
+        defined = {},
+        options = options
+    }
+    local res
+    if not options.forward_reference then
+        res = copy_schema(schema, context)
+    else
+        res =  copy_schema(schema, context)
+        -- Check if all references are resolved.
+        for typeid, cnt in pairs(context.defined) do
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/frontend.lua]: both are fixed
defined -> type_state

File: avro_schema/frontend.lua [hunk 9]
```diff
@@ -548,9 +587,49 @@ copy_schema_error = function(fmt, ...)
...
+        type_non_nullable.nullable = nil
+    end
+end
+
 -- validate schema definition (creates a copy)
 local function create_schema(schema, options)
-    return copy_schema(schema, {scope={}, options=options})
+    local context = {
+        scope = {},
+        -- how many times typeid is defined; see `DEFINED` union
```
[the remark this answers is not in the document]
Comment (Khatskevich) [reply avro_schema/frontend.lua]: deleted

File: avro_schema/frontend.lua (hunk 4 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: added

File: avro_schema/frontend.lua (hunk 3 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: moved

File: avro_schema/frontend.lua (hunk 1 above)
Comment (Totktonada) [reply avro_schema/frontend.lua]: Condition is one the same level as the branch body, so some visual separator is needed. Eigther double indent for condition (I ised this method personally) or `then` on separate line.

File: avro_schema/frontend.lua (hunk 5 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: Do not get the point.

Comment (Totktonada) [reply avro_schema/frontend.lua]: context.type_state[typeid] = {name = ...}
context.type_stateptype_id .. '*'] = {nullable = ...}

The expectation that the field sets is such is not obvious.

Comment (Totktonada) [reply avro_schema/frontend.lua]: Ouch, yep, the original commect was about DEFINED and REFERENCED, that is another thing.

Comment (Totktonada) [reply avro_schema/frontend.lua]: The original comment was about

context.type_state[typeid] = {name = ...}
context.type_stateptype_id .. '*'] = {nullable = ...}

for DEFINED and {} for REFERENCED.

File: avro_schema/frontend.lua (hunk 4 above)
Comment (Totktonada) [reply avro_schema/frontend.lua]: thx

File: avro_schema/frontend.lua [hunk 10]
```diff
@@ -135,6 +142,57 @@ local copy_schema
...
+end
+
+-- get type
+local function scope_get_type(context, typeid, nullable)
+    local scope = context.scope
+    local _, typeid = extract_nullable(typeid)
+    local full_type = nullable and typeid .. "*" or typeid
+    if scope[full_type] then
+        assert(context.type_state[typeid] == TYPE_STATE.REFERENCED or
+        context.type_state[typeid] == TYPE_STATE.DEFINED)
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Indent.

File: avro_schema/frontend.lua [hunk 11]
```diff
@@ -342,8 +401,8 @@ copy_schema = function(schema, context, ns, open_rec)
                             insert(path, res.fields[ptr].name)
                         end
                         error(format('Record %s contains itself via %s',
-                                     field.type.name,
-                                     concat(path, '/')), 0)
+                        field.type.name,
+                        concat(path, '/')), 0)
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: Indent.

File: avro_schema/frontend.lua [hunk 12]
```diff
@@ -378,13 +437,12 @@ copy_schema = function(schema, context, ns, open_rec)
                 end
                 dcache[res] = fieldmap
                 open_rec[res] = nil
-                return res
+                return scope_get_type(context, name, nullable)
             elseif xtype == 'enum' then
-                res = { type = 'enum', symbols = {}, nullable = nullable }
+                res = { type = 'enum', symbols = {} }
```
Comment (Totktonada) [inline avro_schema/frontend.lua]: But you leave 'nullable = nullable' fix array and fixed below. Why?

File: avro_schema/frontend.lua (hunk 10 above)
Comment (Khatskevich) [reply avro_schema/frontend.lua]: fiexd

**Pull request merged**